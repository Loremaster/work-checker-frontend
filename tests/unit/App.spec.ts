import { mount, createLocalVue } from "@vue/test-utils";
import Vuetify from "vuetify";
import Vue from "vue";
import Vuex from "vuex";
import flushPromises from "flush-promises";
import Component from "@/App.vue";

const localVue = createLocalVue();

localVue.use(Vuex);

Vue.use(Vuetify);

let actions: any;
let getters: any;
let store: any;
let wrapper: any;

describe("App.vue", () => {
  beforeEach(() => {
    actions = { setToken: jest.fn(), setUserRole: jest.fn(), signOut: jest.fn() };
    getters = { signedIn: jest.fn() };
  });

  const setup = async () => {
    store = new Vuex.Store({ actions, getters });
    wrapper = mount(Component, {
      store,
      localVue,
      stubs: ["notifications", "router-link", "router-view"],
    });

    await flushPromises();

    return wrapper;
  };

  it("renders correctly", async () => {
    await setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("sets token on page load", async () => {
    await setup();

    expect(actions.setToken).toHaveBeenCalled();
  });

  it("sets user's role on page load", async () => {
    await setup();

    expect(actions.setUserRole).toHaveBeenCalled();
  });

  describe("user is not signed in", () => {
    it("renders button to sign in", async () => {
      getters = { signedIn: jest.fn().mockReturnValue(false) };
      await setup();
      const button = wrapper.findAll(".v-btn__content").at(1);

      expect(button.text()).toEqual("Sign in");
    });
  });

  describe("user signed in", () => {
    beforeEach(async () => {
      getters = { signedIn: jest.fn().mockReturnValue(true) };
      await setup();
    });

    it("renders button the the list of users", () => {
      const button = wrapper.findAll(".v-btn__content").at(1);

      expect(button.text()).toEqual("Users");
    });

    it("let user sign out", () => {
      const button = wrapper.findAll(".v-btn__content").at(2);

      expect(button.text()).toEqual("Sign out");

      button.trigger("click");

      expect(actions.signOut).toHaveBeenCalled();
    });
  });
});
