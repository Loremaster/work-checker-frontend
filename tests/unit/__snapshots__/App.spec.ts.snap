// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App.vue renders correctly 1`] = `
<div data-app="true" class="application theme--light" id="app">
  <div class="application--wrap">
    <nav class="v-toolbar v-toolbar--fixed theme--light" style="margin-top: 0px; padding-right: 0px; padding-left: 0px; transform: translateY(0px);">
      <div class="v-toolbar__content" style="height: 64px;">
        <div class="v-toolbar__title headline text-uppercase"><span>VV</span></div>
        <div class="spacer"></div>
        <div class="v-toolbar__items">
          <router-link-stub class="v-btn v-btn--flat v-btn--router theme--light">
            <div class="v-btn__content">Home</div>
          </router-link-stub>
          <!---->
          <!----> <button type="button" class="v-btn v-btn--flat theme--light">
            <div class="v-btn__content">Sign out</div>
          </button>
        </div>
      </div>
    </nav>
    <main class="v-content" style="padding: 64px 0px 0px 0px;">
      <div class="v-content__wrap">
        <notifications-stub></notifications-stub>
        <router-view-stub></router-view-stub>
      </div>
    </main>
  </div>
</div>
`;
